### PR TITLE
DPC-3636: Extract api client as shared engine for dpc-admin, dpc-web, and dpc-portal

### PR DESCRIPTION
**Pull request by jdettmannnava**
## 🎫 Ticket

https://jira.cms.gov/browse/DPC-3636

## 🛠 Changes

Api Client (with helpers FhirResourceBuilder and OrganizationSubmitSerializer) abstracted from dpc-web and dpc-admin into a separate engine to be included in both projects, as well as dpc-portal.

Api Client renamed to DpcClient to make more clear what it is, especially given the new PECOS client on the horizon.

## ℹ️ Context for reviewers

The bulk of the code for this review is generated by Rails as part of the engine creation process. The critical pieces of code are:
- engines/api_client/app/services/dpc_client.rb
- engines/api_client/app/services/fhir_resource_builder.rb
- engines/api_client/spec/services/dpc_client_spec.rb
- {dpc-admin,dpc-web}/app/models/registered_organization.rb

~~We don't have a gem server, so we are loading the gem via git. We will need to add another pull request for the api_client alone in order to merge it into master, so we can update the Gemfiles to point to the master branch and have everything still work.~~ (Now loading from path, so this is not longer true.)

## ✅ Acceptance Validation
First, I set the dpc app running. then I got a new golden macaroon.

In the rails console for both dpc-web and dpc-admin, I did the following:

- Updated the API_METADATA_URL and GOLDEN_MACAROON to match the running dpc app
- Created an Organization
- Created a RegisteredOrganization linked to the above Organization and the default fhir endpoint. 
- Validated that the organization was now present in the dpc app database
- Validated I received the appropriate information from the DpcClient#get_organization with the new RegisteredOrganization#api_id
- Updated the Organization and verified that the changes where in the dpc app database
- Updated the RegisteredOrganization (changed "enabled" back and forth) and confirmed that both both update_organization and update_endpoint where called
- Built a new ClientTokenManager with the new RegisteredOrganization
- Created two new client_tokens via the ClientTokenManager
- Retrieved the new client_tokens via the ClientTokenManager
- Deleted one of the new client_tokens via the ClientTokenManager
- Built a new PublicKeyManager
- Followed the instructions for creating a public pem and signature file from dpc-web/app/views/public_views/new.html.erb
- Created a public_key with these parameters via the PublicKeyManager
- Retrieved the public_key via the PublicKeyManager
- Deleted the public_key via the PublicKeyManager (dpc-web only, as this is not available in dpc-admin)

## 🔒 Security Implications

- [x] This PR adds a new software dependency or dependencies.
- [ ] This PR modifies or invalidates one or more of our security controls.
- [ ] This PR stores or transmits data that was not stored or transmitted before.
- [ ] This PR requires additional review of its security implications for other reasons.

If any security implications apply, add Jason Ashbaugh (GitHub username: StewGoin) as a reviewer and do not merge this PR without his approval.

### CODE REVIEW

File: dpc-portal/Rakefile
Comment (jdettmannnava) [inline dpc-portal/Rakefile:1]: I was having difficulty getting portal to build with the new gem, so I tried to fix up webpacker by copying this from web-admin.

File: dpc-portal/lib/tasks/webpacker.rake
Comment (jdettmannnava) [inline dpc-portal/lib/tasks/webpacker.rake:1]: I was having difficulty getting portal to build with the new gem, so I tried to fix up webpacker by copying this from web-admin.

File: engines/api_client/spec/fixtures/files/stubbed_key.pem
Comment (MEspositoE14s) [inline engines/api_client/spec/fixtures/files/stubbed_key.pem:1]: Can we get into any trouble committing this?  Whenever I see one of these it makes me nervous.

File: dpc-api-client-test.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+set -e
+
+echo "┌───────────────────────┐"
+echo "│                       │"
+echo "│ Running Api Gem Tests |"
+echo "│                       │"
+echo "└───────────────────────┘"
+
+# Build the container
+docker build -f engines/api_client/Dockerfile engines/api_client -t api-client
```
Comment (MEspositoE14s) [inline dpc-api-client-test.sh:11]: I tried to test this out, but the Dockerfile is missing in `engines/api_client`

Comment (jdettmannnava) [reply dpc-api-client-test.sh:11]: Added!

File: engines/api_client/spec/fixtures/files/stubbed_key.pem
Comment (jdettmannnava) [reply engines/api_client/spec/fixtures/files/stubbed_key.pem:1]: I copied it from dpc-admin, so I think it should be OK?

Comment (kyeah) [reply engines/api_client/spec/fixtures/files/stubbed_key.pem:1]: yeah if it's just a stub/fake key and only for tests it's definitely cool 👋

**Review by kyeah — APPROVED** [review]
Thank you! Didn't walk through all of your manual testing steps, but it seems like it was thorough. It would be good to verify the web and web-admin functionality directly through the UI but I'm not sure that's possible locally (I tried to log into the Admin portal after setting Github Oauth using the secrets in the dev environment, but it errors out when redirecting back to the localhost domain.)

It might be a good idea to deploy this branch to the dev environment to test things out before we merge this in.

**Comment by jdettmannnava** [comment]
> Thank you! Didn't walk through all of your manual testing steps, but it seems like it was thorough. It would be good to verify the web and web-admin functionality directly through the UI but I'm not sure that's possible locally (I tried to log into the Admin portal after setting Github Oauth using the secrets in the dev environment, but it errors out when redirecting back to the localhost domain.)
> 
> It might be a good idea to deploy this branch to the dev environment to test things out before we merge this in.

1. Should we make a ticket to get admin running locally?
2. Are there instructions on how to deploy to dev and testing credentials?

**Comment by kyeah** [comment]
> Should we make a ticket to get admin running locally?

Good question! I think let's keep it as is unless we find that we keep coming back and making changes to that application — then we can definitely prioritize improving the developer experience there.

File: package-lock.json [hunk 2]
```diff
@@ -1,1867 +0,0 @@
-{
```
[the remark this answers is not in the document]
Comment (jdettmannnava) [reply package-lock.json:1]: Restored!

File: dpc-portal/Gemfile.lock [hunk 3]
```diff
@@ -1,3 +1,15 @@
+GIT
+  remote: https://github.com/CMSgov/dpc-app.git
+  revision: d85a3962090a4dd3ddf4137a69694f532842a32d
+  branch: jd/dpc-api-client-to-dpc-portal
```
[the remark this answers is not in the document]
Comment (jdettmannnava) [reply dpc-portal/Gemfile.lock]: Updated

File: Makefile [hunk 4]
```diff
@@ -19,7 +19,7 @@ admin:
```
[the remark this answers is not in the document]
Comment (jdettmannnava) [reply Makefile:19]: Updated

File: dpc-web/config/environments/test.rb [hunk 5]
```diff
@@ -1,5 +1,6 @@
 Rails.application.configure do
   # Settings specified here will take precedence over those in config/application.rb.
+  config.active_record.dump_schema_after_migration = false
```
[the remark this answers is not in the document]
Comment (jdettmannnava) [reply dpc-web/config/environments/test.rb]: Restored

File: dpc-admin/config/initializers/lograge.rb [hunk 6]
```diff
@@ -2,7 +2,11 @@
...
 Rails.application.configure do
   config.lograge.custom_options = lambda do |event|
-    { params: event.payload[:params],
-      level: event.payload[:level] }
+    { ddsource: 'ruby',
+      params: event.payload[:params].reject { |k| %w(controller action).include? k },
+      environment: ENV['ENV'] || :development,
+      exception: event.payload[:exception],
+      level: ENV['LOG_LEVEL'] || :info
+    }
```
Comment (kyeah) [inline dpc-admin/config/initializers/lograge.rb:10]: Can we include this in dpc-portal too? :D

We should also consider if we need a basic PII scrubber for these exception messages — something we can keep in the backlog as a security / path to production item.

**Review by kyeah — APPROVED** [review]
LGTM after your manual review in dev!

File: dpc-admin/config/initializers/lograge.rb (hunk 6 above)
Comment (jdettmannnava) [reply dpc-admin/config/initializers/lograge.rb:10]: Done!

**Pull request merged**